libs/wai-utilities/src/Network/Wai/Utilities/Server.hs

```diff
@@ -184,21 +166,39 @@ measureRequests m rtree = withPathTemplate rtree $ \p ->
       requestCounter m p . duration 30 12 m p
 {-# INLINEABLE measureRequests #-}
 
+-- | Alias for @'catchErrors'' True@.
+catchErrors  :: Logger -> OnErrorMetrics -> Middleware
+catchErrors = catchErrors' True
+{-# INLINEABLE catchErrors #-}
+
 -- | Create a middleware that catches exceptions and turns
 -- them into appropriate 'Error' responses, thereby logging
 -- as well as counting server errors (i.e. exceptions that
 -- yield 5xx responses).
-catchErrors :: Logger -> Metrics -> Middleware
-catchErrors l m app req k =
-    app req k `catch` errorResponse
+--
+-- The boolean determines whether response values with
+-- status >= 400 are allowed (servant), or whether we should
+-- not expect and consequently warn about them (wai-route).
+catchErrors' :: Bool -> Logger -> OnErrorMetrics -> Middleware
+catchErrors' warnNonExceptFailures l m app req k =
+    app req k' `catch` errorResponse
   where
     errorResponse ex = do
         er <- runHandlers ex errorHandlers
         when (statusCode (Error.code er) >= 500) $
-            logIO l Log.Error (Just req) (oneline <$> show ex)
+            logIO l Log.Error (Just req) (show ex)
         onError l m req k er
-    oneline c = if isSpace c then ' ' else c
-{-# INLINEABLE catchErrors #-}
+
+    -- catch status >= 400 even if it does not come in the form of an exception, but as a
+    -- response, and log the request id.  not much else we can do about it without taking
+    -- apart the streaming response.  this should be fixed whenever it is found in the logs.
+    k' resp = do
+        when (warnNonExceptFailures && statusCode (responseStatus resp) >= 400) $ do
+            Log.warn l
+                $ field "request" (fromMaybe "N/A" (lookupRequestId req))
```

Comment:
Maybe `Unknown` is clearer than `N/A`

Reply:
I would like to keep this because (a) I disagree, and (b) I'm not sure which parts of out tool landscape depend on this.  How important is this to you?

Reply:
Not very; that's fine :)